[kaikquah] IP

kaikquah opened the pull request:
# Gale 💨 
>“Your mind is for having ideas, not holding them.” – David Allen

## About
Gale frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- FAST SUPER FAST to use

All you need to do is,

1. Click it from [here](https://github.com/kaikquah/ip)
2. double-click it.
3. add your tasks, and
4. let it manage your tasks for you! 😎 

And it is **FREE!**

## Progress
- [x] Week 2
- [x] Week 3
- [x] Week 4
- [ ] Week 5

## Code Snippet
Here's a look of the `main` method from `src/main/java/gale/Gale.java`:
```java
public static void main(String[] args) {
        Gale gale = new Gale();
        gale.run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Level-0

Commit:
Level 1 commit. Made changes to add personality and separated methods.

Commit:
Level 2 commit. Changed name for the task-storing chatbot and added a list of tasks.

Commit:
Level 3 commit. Added new Task class. Edited Gale.

Commit:
Level 4 commit. Added subclasses of Task: Deadline, Event, ToDo

Commit:
text-ui-test commit

Commit:
Level-5 commit. Exceptions added along with GaleException class

Commit:
Level-6. Delete added

Commit:
Add Save feature in Gale class

A Storage class was added to bundle the file-saving code,
and Task class was made abstract so that I could use
polymorphism in the subclasses of the Task class.

Commit:
Merge branch 'branch-Level-7'

Commit:
Add date and time logic

Event and Deadline classes were implemented with date and time logic.
User will have to add date and time when they specify their event
and deadline tasks.

Commit:
Merge branch 'branch-Level-8'

Commit:
Extract out related code as classes

Makes the code follow more of OOP standards.
Added Parser, TaskList, Ui classes.

Commit:
Package all files into gale

Added all java files into the package gale

Commit:
Merge branch 'add-gradle-support'

Commit:
Add Gradle support and JUnit tests

Added JUnit tests in the form of GaleTest and ParserTest classes.
Used Gradle to automate the running and testing of my application.

Commit:
Add JavaDoc comments

JavaDoc comments were added to the classes of the main application folder.

Commit:
Follow the coding standard

Commit:
Add find task feature

This gives users a way to find a task by searching for a keyword in the task description.

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-Level-9'

Commit:
Merge branch 'branch-A-CodingStandard'

anselmlong commented on build.gradle:
I think it may be best practice to add this to your .gitignore file!

anselmlong commented on gradlew:
Same here, you might want to .gitignore this.

anselmlong commented on src/main/java/gale/Deadline.java:
```diff
@@ -0,0 +1,42 @@
+ *
+ * @author kaikquah
+ */
+public class Deadline extends Task {
```
Would it be better to suffix all your task files with "-Task"? For example, "DeadlineTask", or "ToDoTask" so your files are clearly separated?

anselmlong commented on src/main/java/gale/Deadline.java:
```diff
@@ -0,0 +1,42 @@
+ */
+public class Deadline extends Task {
+    private LocalDateTime by;
+    private DateTimeFormatter outputFormatter = DateTimeFormatter.ofPattern("MMM d yyyy HH:mm");
```
Good job on abstracting this out!

anselmlong commented on src/main/java/gale/Event.java:
```diff
@@ -0,0 +1,50 @@
+package gale;
```
You could consider further separating your packages into `gale.tasks` and other items!

anselmlong commented on src/main/java/gale/Gale.java:
```diff
@@ -0,0 +1,158 @@
+     * Creates a new Gale instance with the default file path to store the output.
+     * <p>This constructor sets up the Ui and Storage instances, and loads the existing TaskList.</p>
+     */
+    public Gale() {
```
Clean code!

anselmlong commented on src/main/java/gale/Gale.java:
```diff
@@ -0,0 +1,158 @@
+     *
+     * @param filePath where the output file is stored
+     */
+    public Gale(String filePath) {
```
I love how you overloaded the constructor! 👍

anselmlong commented on src/main/java/gale/Gale.java:
```diff
@@ -0,0 +1,158 @@
+     */
+    protected void saveTasks() {
+        try {
+            storage.saveTasks(taskList.getTaskList());
```
Good tell don't ask here.

anselmlong commented on src/main/java/gale/Gale.java:
```diff
@@ -0,0 +1,158 @@
+            String input;
+            try {
+                input = scanner.nextLine().trim();
+                if (input.equalsIgnoreCase("bye")) {
```
Would using a switch-case here be more effective?

anselmlong commented on src/main/java/gale/Parser.java:
```diff
@@ -0,0 +1,135 @@
+     * @param input the input string to be parsed
+     * @return the Task object parsed from the input string
+     * @throws GaleException if the input string does not match any of the task types
+     */
```
Good, descriptive Javadocs here.

anselmlong commented on src/main/java/gale/Parser.java:
```diff
@@ -0,0 +1,135 @@
+        if (input.length() <= 9) {
+            throw new GaleException("Your deadline got tossed by the wind! Please use 'deadline [description] /by [date]'.");
+        }
+        String[] strA  = input.substring(9).split("/by");
```
Maybe you could use a more descriptive variable name here, like inputSplit?

anselmlong commented on src/main/java/gale/Parser.java:
```diff
@@ -0,0 +1,135 @@
+        if (input.length() <= 6) {
+            throw new GaleException("Your event is lost in the wind! Please use 'event [description] /from [start] /to [end]'.");
+        }
+        String[] strA = input.substring(6).split("/from|/to");
```
See previous comment!

anselmlong commented on src/main/java/gale/Storage.java:
```diff
@@ -0,0 +1,108 @@
+        boolean isDone = parts[1].equals("1");
+        String desc = parts[2];
+        try {
+            switch (type) {
```
Good job following Switch-Case formatting.

anselmlong commented on src/main/java/gale/Storage.java:
```diff
@@ -0,0 +1,108 @@
+        } catch (DateTimeParseException e) {
+            return null;
+        }
+        if (isDone) task.markAsDone();
```
Unfortunately, this violates the coding standards for if statements 😢 .
Refer to [this](https://se-education.org/guides/conventions/java/intermediate.html#conditionals)!

anselmlong commented on src/main/java/gale/Ui.java:
```diff
@@ -0,0 +1,130 @@
+    public void printAddedTask(Task task, int taskCount) {
+        System.out.println(HORIZONTAL_LINE);
+        System.out.println("Whoosh! Task \"" + task + "\" added to my windy memory.");
+        System.out.println("Now you have " + taskCount + " task" + (taskCount == 1 ? "" : "s")
```
I like how you account for the plural form (I didn't...).

anselmlong commented on src/main/java/gale/Ui.java:
```diff
@@ -0,0 +1,130 @@
+    public void showFoundTasks(ArrayList<Task> foundTasks, String keyword) {
+        System.out.println(HORIZONTAL_LINE);
+        if (foundTasks.isEmpty()) {
+            System.out.println("Oops. I combed through the clouds, but there were no tasks found with that keyword.");
```
Cute!

anselmlong commented on src/test/java/gale/GaleTest.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    @Test
+    public void testAddTask_success() throws GaleException {
```
Unfortunately again, this doesn't follow test case naming standards!
> Underscores may be used in test method names using the following three part format featureUnderTest_testScenario_expectedBehavior()

> e.g. sortList_emptyList_exceptionThrown() getMember_memberNotFound_nullReturned

More details [here!](https://se-education.org/guides/conventions/java/intermediate.html#naming)

anselmlong replied:
For this, it could be addToDoTask_validInput_successMessage()?

devivy replied:
same review; 

doesn't follow test case naming standards:(
for help [click here](https://se-education.org/guides/conventions/java/intermediate.html#naming)

the correct format could be addToDoTask_validInput_successMessage() (maybe)

kaikquah replied:
Thanks for the help, will edit it!

anselmlong commented on src/test/java/gale/ParserTest.java:
```diff
@@ -0,0 +1,36 @@
+public class ParserTest {
+
+    @Test
+    public void taskCommand_wrongCommandFormat_exceptionThrown() {
```
You got the naming right, but I'm not sure whether the name here is very intuitive! It seems like you're trying to test your parser, so I'm a bit confused on what `taskCommand` is.

devivy replied:
naming is not very clear, a bit confusing

kaikquah replied:
Will edit it, thanks for the help guys!

anselmlong reviewed (commented):
Great work! 💯 Overall, just some naming convention changes. Honestly, I'm quite blown away by your code :)

Commit:
Add checkstyle to check for coding standards

Commit:
Merge branch 'branch-A-Checkstyle'

devivy commented on src/main/java/gale/Deadline.java:
everything looks clean:) 

whats good:
- smart use of abstraction 
- all the instructions followed to T 

you could have further 
- separated your packages into gale.tasks etc
- made your class names for specific ( DeadlineTask, ToDoTask,etc)

devivy commented on src/main/java/gale/Event.java:
neat!

devivy commented on src/main/java/gale/Gale.java:
everything looks clean! 😊

what's good:
-smart use of abstraction
- good exception handling
- clear JavaDoc comments
- use of constructor overloading 
- use of Tell Don't Ask


you could have further:
- separated packages (e.g., gale.tasks, gale.exceptions)
- considered using enums for task types
- used a switch-case( maybe more effective)
- implemented a command pattern for user inputs

devivy commented on src/main/java/gale/Parser.java:
what's good:
- flexible date-time parsing with multiple formatters
- detailed error messages 
- descriptive java docs
- clear separation of parsing logic for different task types

code is perfect:)

devivy commented on src/main/java/gale/Storage.java:
perfect!!!

devivy commented on src/main/java/gale/Task.java:
solid implementation!!

what's good:
- clear abstraction with abstract class and methods
- comprehensive javadoc comments
- good encapsulation with private fields and public methods

you could have further:
- used final keyword for immutable fields (e.g., description)
- considered using an enum for task status

devivy commented on src/main/java/gale/TaskList.java:
perfect:)

devivy commented on src/main/java/gale/Ui.java:
clean! 

what's good:
- clear separation of ui concerns
- good use of string formatting for plural/singular task counts

devivy reviewed (commented):
overall, really really good work:)
nice theme

Commit:
Add the respective classes to their packages

Commit:
Add GUI for Gale using JavaFX

Preliminary round of using JavaFX, will switch to FXML later

Commit:
Merge branch 'branch-Level-10'

Commit:
Add assertions to the main logic flow

Assertions are not added yet, thus we are not able
to catch any illogical flow in the program.

Let's add assertions to parts of the program
where it is important for the program to run
smoothly.

In the run() method, we assert that the ui is not null.
In the getResponse method, we assert that the input
is not null and instead a String.
In the saveTasks method in the Storage class, we assert
that taskList is not null.

Commit:
Improve the code quality

Code is not well factored yet as certain guard
clauses could be extracted out.
Additionally, some common String messages
could be extracted out and reused.

Let's add more guard clauses to make the
happy path of the code more prominent.
Additionally, let's extract out the common parts
of the code that could be reused again to avoid
tardiness.

kaikquah committed:
Merge pull request #1 from kaikquah/branch-A-Assertions

Add assertions to the main logic flow

Commit:
Merge branch 'master' into branch-A-CodeQuality

kaikquah committed:
Merge pull request #2 from kaikquah/branch-A-CodeQuality

Improve the code quality

Commit:
Add a way for Priority to be stored in each task

We are lacking a way to add priority for each task, but it would be good
for the user to indicate the priority for each of the tasks that he has.
However, in the case that the user does not want to add a priority to
the task, he would also be allowed to do so.

Let's create a Priority class that encapsulates the priority of each
task, and update the Task class and its subclasses to be able to store
and display the priority of each individual task when each command is
executed.

Commit:
Merge branch 'branch-C-Priority'

Commit:
Add FXML and CSS enhancements to the GUI

Commit:
Merge branch 'branch-Level-10'

Commit:
Fix code quality in the Gale class

Certain methods in the Gale class could be further abstracted out.
The methods are: findTasks, handleTaskMarking and deleteTask.

Let's create new helper methods to abstract out the logic within
the aforementioned methods and place them in their respective
classes.

In the Parser class, the parseIndex and parseKeyword methods were
created.

In the TaskList class, the markTask and deleteTask methods were
edited.

Commit:
Add background image for GUI

Commit:
Update JUnit test to account for priority of each task

Commit:
Stop tracking build.gradle

Commit:
Improve code quality for Parser and Ui classes

Commit:
Fix bug related to parsing of Event tasks

Commit:
Add and update JUnit tests

Commit:
Add User guide for Gale

Commit:
Change product screenshot

Commit:
Update JUnit Test

Commit:
Update User guide

Commit:
Update user guide

Commit:
Fix bug related to event parser

Commit:
Update ReadMe

Commit:
Fix bug related to todo event parser

Bug occurred where input "todoexample" would result in an event of
"xample" being added as there was code to remove the first 5
characters of the input string.
Now the bug is fixed by changing up the code to detect if there
is no whitespace between the command todo and the description and
will throw an appropriate exception.
